### DISCUSSION TITLE
Allow the possibility to specify another config directory

### DISCUSSION THREAD

**Discussion by federico-terzi** [espanso/espanso#1173] [Feature requests and Ideas]
The user will place a little config in the default path that will point to another directory containing the actual configuration to load.
This will make it easier to sync configuration files without creating symbolic links

**Comment by ax42** [comment]
Second this -- would allow easy integration (on Mac) with dotfile management systems like Homeshick.  'The unix way' (which also works for MacOS) would be to have configs in `~/.espanso` or `~/.config/espanso`.

**Comment by MuhammedZakir** [comment]
> [...] The unix way' (which also works for MacOS) would be to have configs in `~/.espanso` or `~/.config/espanso`.

+1. `~/.config/espanso` is better than the current one IMHO. It would make switching between different *nix OS easier.

Edit: How about allowing config dir to be set via ENV variable?

**Comment by federico-terzi** [comment]
@MuhammedZakir funny thing is, the first version of espanso used `~/.espanso` as default directory and many people complained about that :)

> Edit: How about allowing config dir to be set via ENV variable?

That could be a good idea, even though on some system they might be difficult to set (keep in mind that espanso is managed by Systemd on Linux and Launchd on macOS, which handle env variables differently)

**Comment by MuhammedZakir** [comment]
> @MuhammedZakir funny thing is, the first version of espanso used `~/.espanso` as default directory and many people complained about that :)

Ah! I did a quick search and found #73 which has been closed by a PR. Would have been nice all fallback dirs are documented somewhere or did I miss it? I guess *nix users' problem is solved already?

> 
> > Edit: How about allowing config dir to be set via ENV variable?
> 
> That could be a good idea, even though on some system they might be difficult to set (keep in mind that espanso is managed by Systemd on Linux and Launchd on macOS, which handle env variables differently)

So then we are back to square one, place a config in default path! 😅 Let's go with that. Also, please add a cmd flag for it.

Thanks a lot for this awesome software, btw!

**Comment by ax42** [comment]
Setting config in an ENV variable is something I've seen for server-type implementations (spinning up Docker containers for example), seldom for 'client' software.  The question then becomes 'where do I set the ENV variable', and that is then in another dotfile, and it is unclear whether that would work (because launchd does not necessarily launch a shell which will parse a dotfile).

+1 for a command flag (as well as a default location).

**Comment by federico-terzi** [comment]
Hey guys,
I'll have to think carefully about the best way to implement the command line flag because as of now, it might not be trivial. Keep in mind that espanso is currently set up in a multi-process architecture, and the way those processes are spawned depends on the underlying operating system, each one of them with its quirks.
That said, I'll keep an eye on it :)

**Comment by safakgur** [comment]
Hi @federico-terzi - thanks for espanso (which has been indispensable for me), and sorry for resurrecting this topic. I just wanted to get your opinion about something.

As linking is the only way to specify a config directory right now, I've been using the following PowerShell script whenever I setup a new Windows PC:

```powershell
New-Item -Type Junction -Path "$env:AppData\espanso" -Value "$env:OneDriveConsumer\Apps\Espanso"
winget install Espanso.Espanso
```

...which creates the junction first (assuming the app will use `AppData\Roaming\espanso` to store its config), and then installs the app.

I opted to create the junction before installing espanso because I wanted to make sure the espanso folder isn't there yet - otherwise, I thought, I'd have to manually delete the original folder first in order to create the junction. So, here are a few questions:

1. Can I trust espanso will always use `%APPDATA%\espanso` folder to store config on Windows, or would it be safer if I install the app first and use whatever path `espanso path config` returns for the junction?

2. Also, would installing the app and running `espanso path config` (without ever touching `espansod.exe`) cause the config folder to be created?

   If not, I guess I can go with `install -> get-path -> create-junction`. But if they do (or can in the future) create the initial config folder, then my script will have to do `install -> get-path -> *remove-folder* -> create-junction`.

3. I currently use OneDrive to sync espanso config in both my Windows and Mac devices. As far as I can tell, the config doesn't really have anything platform-specific. But is this a use case you explicitly support, or do you think it's better if I keep them separate?

Thank you!

**Comment by ptrkhh** [comment]
You could release two types of app e.g Espanso and Espanso Portable. A lot of other apps are distributed this way to account for different people's needs

**Comment by saiqulhaq-hh** [comment]
in my windows 11 PC, the config file is located in C:\Users\xxx\app\xxx\xxx\xx
so is it not possible to move the YML file to Google Drive?
i tried to use a Shortcut file, but it doesn't work
Espanso app creates a new base.yml file, not reading the existing shortcut file

**Reply by smeech** [reply]
The only working method is to use symlinks, as described [here](https://espanso.org/docs/sync/).
